doc/read-the-docs-site/howtos/asdata.rst

```diff
@@ -0,0 +1,89 @@
+The ``asData`` function takes the definition of a data type and replaces it with an equivalent definition whose representation uses ``Data`` directly.
+
+For example, if we wanted to use it on the types from the :ref:`auction example<simple_example>`, we would put the datatype declarations inside a Template Haskell quote and call ``asData`` on it.
+
+.. literalinclude:: ../tutorials/AuctionValidator.hs
+   :start-after: BLOCK9
+   :end-before: BLOCK10
+
+This is normal Template Haskell that just generates new Haskell source, so you can see the code that it generates with `{-# OPTIONS_GHC -ddump-splices #-}`, but it will look something like this:
+
+.. code-block::
+
+        PlutusTx.asData
+        [d| data Bid'
+                = Bid' {bBidder' :: PubKeyHash, bAmount' :: Lovelace}
+                deriving newtype (Eq, Ord, ToBuitinData, FromBuiltinData, UnsafeFromBuiltinData)
+            data AuctionRedeemer' = NewBid' Bid | Payout'
+                deriving newtype (Eq, Ord, ToBuitinData, FromBuiltinData, UnsafeFromBuiltinData) |]
+
+        ======>
+
+        newtype Bid' = Bid'2 BuiltinData
+        deriving newtype (Eq, Ord, PlutusTx.ToData, FromData, UnsafeFromData)
+
+        {-# COMPLETE Bid' #-}
+        pattern Bid' :: PubKeyHash -> Lovelace -> Bid'
+        pattern Bid' ...
+
+        newtype AuctionRedeemer' = AuctionRedeemer'2 BuiltinData
+        deriving newtype (Eq, Ord, PlutusTx.ToData, FromData, UnsafeFromData)
+
+        {-# COMPLETE NewBid', Payout' #-}
+        pattern NewBid' :: Bid -> AuctionRedeemer'
+        pattern NewBid' ...
+        pattern Payout' :: AuctionRedeemer'
+        pattern Payout' ...
+
+That is:
+
+- It creates a newtype wrapper around ``BuiltinData``
+- It creates pattern synonyms corresponding to each of the constructors you wrote
+
+This lets you write code "as if" you were using the original declaration that you wrote, while in fact the pattern synonyms are handling conversion to/from ``Data`` for you.
+But any values of this type actually are represented with ``Data``.
+That means that when we newtype-derive the instances for converting to and from ``Data`` we get the instances for ``BuiltinData`` - which are free!
+
+Caveats
+-------
+
+The most important caveat to using ``asData`` is that ``Data`` objects encoding datatypes must also encode the *fields* ``Data``.
+However, when using the pattern synonyms they try to give you exactly what you asked for, which might mean having the fields *not* encoded as ``Data``.
+
+For example, in the ``Bid`` case above the ``Lovelace`` field is represented as a Plutus Core builtin integer.
+However, in order for it to be encoded into the ``Bid`` structure, we need to encode it into ``Data``.
+That means that when you construct a ``Bid`` object you do an ``Integer`` to ``Data`` conversion, and when you pattern match on a ``Bid`` object you do a ``Data`` to ``Integer`` conversion.
```

Comment:
I don't think `Integer` is a very good example, since there's almost certainly no need to bother with a Data-encoded Integer type, vs. just the plain Integer.

Reply:
I'm confused. The point is that you're data-encoding `Bid`, and that _forces_ you to data-encode its `Integer` fields (which is cheap, but not free)